src/super_gradients/training/losses/yolox_loss.py

```diff
@@ -0,0 +1,468 @@
+            -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        L = L_objectivness + L_iou + L_classification + 1[no_aug_epoch]*L_l1
+        where:
+            * L_boxes, L_classification and L_l1 are calculated only between cells and targets that suit them;
+            * L_objectivness is calculated for all cells.
+
+        L_classification:
+            for cells that have suitable ground truths in their grid locations add BCEs
+            to force a prediction of IoU with a GT in a multi-label way
+            Coef: 1.
+        L_iou:
+            for cells that have suitable ground truths in their grid locations
+            add (1 - IoU^2), IoU between a predicted box and each GT box, force maximum IoU
+            Coef: 1.
+        L_l1:
+            for cells that have suitable ground truths in their grid locations
+            l1 distance between the logits and GTs in “logits” format (the inverse of “logits to predictions” ops)
+            Coef: 1[no_aug_epoch]
+        L_objectness:
+            for each cell add BCE with a label of 1 if there is GT assigned to the cell
+            Coef: 1
+
+        :param predictions:     output from all Yolo levels, each of shape
+                                [Batch x Num_Anchors x GridSizeY x GridSizeX x (4 + 1 + Num_classes)]
+        :param targets:         [Num_targets x (4 + 2)], values on dim 1 are: image id in a batch, class, box x y w h
+
+        :return:                loss, all losses separately in a detached tensor
+        """
+        x_shifts, y_shifts, expanded_strides, transformed_outputs, raw_outputs = self.prepare_predictions(predictions)
+
+        bbox_preds = transformed_outputs[:, :, :4]  # [batch, n_anchors_all, 4]
+        obj_preds = transformed_outputs[:, :, 4:5]  # [batch, n_anchors_all, 1]
+        cls_preds = transformed_outputs[:, :, 5:]  # [batch, n_anchors_all, n_cls]
+
+        # calculate targets
+        total_num_anchors = transformed_outputs.shape[1]
+        cls_targets = []
+        reg_targets = []
+        l1_targets = []
+        obj_targets = []
+        fg_masks = []
+
+        num_fg, num_gts = 0., 0.
+
+        for batch_idx in range(transformed_outputs.shape[0]):
+            labels_im = targets[targets[:, 0] == batch_idx]
+            num_gt = labels_im.shape[0]
+            num_gts += num_gt
+            if num_gt == 0:
+                cls_target = transformed_outputs.new_zeros((0, self.num_classes))
+                reg_target = transformed_outputs.new_zeros((0, 4))
+                l1_target = transformed_outputs.new_zeros((0, 4))
+                obj_target = transformed_outputs.new_zeros((total_num_anchors, 1))
+                fg_mask = transformed_outputs.new_zeros(total_num_anchors).bool()
+            else:
+                # GT boxes to image coordinates
+                gt_bboxes_per_image = labels_im[:, 2:6].clone()
+                # gt_bboxes_per_image = labels_im[:, 2:6] * self.im_size
```

Comment:
is this a comment or should be deleted?

Reply:
deleted (used to have unnormalised labels earlier on...)